src/Lib/BSP/IO/BspFileReader.cs

```diff
@@ -0,0 +1,210 @@
+
+                lump.Type = type;
+                lumpHeader.Offset = ReadInt32();
+                int length = ReadInt32();
+                lump.Version = ReadInt32();
+                int fourCc = ReadInt32();
+                if (fourCc == 0)
+                {
+                    lumpHeader.CompressedLength = -1;
+                    lumpHeader.UncompressedLength = length;
+                }
+                else
+                {
+                    lumpHeader.CompressedLength = length;
+                    lumpHeader.UncompressedLength = fourCc;
+                }
+
+                Console.WriteLine($"Lump {type}({(int)type})"
+                                  + $"\toffset: {lumpHeader.Offset}"
+                                  + $"\t length: {length}"
+                                  + $"\t Version: {lump.Version}"
+                                  + $"\t FourCc: {fourCc}");
+
+                _bsp.Lumps.Add(type, lump);
+                Lumps.Add(new Tuple<Lump, LumpHeader>(lump, lumpHeader));
+            }
+
+            _bsp.Revision = ReadInt32();
+
+            UpdateGameLumpLength();
+
+            SortLumps();
+
+            if (CheckOverlapping())
+                throw new InvalidDataException("overlapping lumps bad");
+        }
+
+        //finding the real gamelump length by looking at the next lump
+        private void UpdateGameLumpLength()
+        {
+            Lump gameLump = null;
+            LumpHeader gameLumpHeader = null;
+            foreach (var l in Lumps.OrderBy(x => x.Item2.Offset))
+            {
+                Lump lump = l.Item1;
+                LumpHeader header = l.Item2;
+                if (lump is GameLump)
+                {
+                    gameLump = lump;
+                    gameLumpHeader = header;
+                }
+                else if (gameLump is not null && header.Offset != 0 && header.Offset != gameLumpHeader.Offset)
+                {
+                    gameLumpHeader.UncompressedLength = header.Offset - gameLumpHeader.Offset;
+                    Console.WriteLine($"changed gamelump length to {gameLumpHeader.Length}");
+                    break;
+                }
+            }
+        }
+
+        //sort by offset so the output file looks more like the input
+        private void SortLumps()
+        {
+            Dictionary<BspLumpType, Lump<BspLumpType>> newLumps = new();
+            foreach (var l in Lumps.OrderBy(x => x.Item2.Offset))
+            {
+                var temp = _bsp.Lumps.Where(x => x.Value == l.Item1).First();
+                newLumps.Add(temp.Key, temp.Value);
+                _bsp.Lumps.Remove(temp.Key);
+            }
+            if (_bsp.Lumps.Any())
+                throw new InvalidDataException("bsp lumps and reader headers didn't match!");
+            _bsp.Lumps = newLumps;
+
+        }
+
+        //for testing
+        private bool CheckOverlapping()
+        {
+            bool ret = false;
```

Comment:
It's a nit I know, but I'd prefer we be more consistent with `var` vs explicit types. For something that's so clearly a `bool` (from the RHS) I think we should `var` in these cases.